### ISSUE TITLE
mTLS

### ISSUE THREAD

**Issue by mkohns** [pomerium/pomerium#534]
**This is a question, no bug or feature requests**

Beside identity-awareness through IDP flow, a mTLS (client certificate) could even be a security layer above. 

Is it possible to configure pomerium to check https request for provided client certificates and check them if trusted?

Scenario:
-> configure pomerium with i.e. public root CA 
-> user generates client certificates (signed by above root CA) 
-> client (browser, curl, whatever) uses mTLS flow to provide client certs in order to establish https connection.

**Comment by desimone** [comment]
@mkohns 

> Is it possible to configure pomerium to check https request for provided client certificates and check them if trusted?

Yes. And it is encouraged.

https://www.pomerium.io/configuration/#policy

See `TLS Client Certificate`. 

https://github.com/pomerium/examples/tree/master/mutual-tls

**Comment by mkohns** [comment]
Great! Thanks I will test and report back!

**Comment by desimone** [comment]
Let me know. Also, would love to add a "recipe" for mTLS if that's something you'd be willing to contribute. https://www.pomerium.io/recipes/ 👍

**Comment by mkohns** [comment]
@desimone 
of course! Would like to contribute to your great project!

**Comment by mkohns** [comment]
@desimone 
I digged deeper into the current mTLS features of pomerium. Here are my thoughts:

1) Currently pomerium supports upstream mTLS configuration. Calls made from pomerium towards upstream container can be secured by mTLS. Here an example:

```
policy:
  - from: https://mtls.example.com
    to: https://mtlsserver:8443
    # use this in case the mtlsserver certificate issuer (ca)
    # is not in the truststore
    tls_custom_ca_file: /pomerium/ca.pem
    # this is the client certificate needed by the upstream mtlsserver
    tls_client_cert_file: /pomerium/client.pem
    # this is the correspoinding key to above client certificate
    tls_client_key_file: /pomerium/client.key
```

The upstream container (mtlsserver) in this example is separately configured in two ways
* serves content by https (server certificate and corresponding key are configured)
* asks clients for client certificates (client certificate CA is configured. The server checks if the provided client certificate is signed by the configured CA)

NOTE: The CA for the communication channel (https) does not necessarily be the same as the client certificate CA. I would like to point this out here. Scenarios with one CA for the https server certificates - and one CA for the client certificates work just fine and isolation is supported.

NOTE: All clients using pomeriums endpoint https://mtls.example.com do not need to provide client certificates. All traffic to this endpoint will be made internally (pomerium<->upstream) with the same configured client certificates.

2) pomerium end-user client certificates
Currently (from my point of view) only the pomerium<->upstream communication can be secured with client certificates. I would be great to add this extra layer of security to the pomerium end-user. I explain this with an nginx example

```
server {
        server_name mtls.example.com;
        listen 443 ssl http2 ;
        # This is the https server certificate (i.e. let-encrpyt)
        ssl_certificate /etc/nginx/certs/fullchain.pem;
        # This is the corresponding server certificate key
        ssl_certificate_key /etc/nginx/certs/key.pem;
        # some extra options 
        ssl_dhparam /etc/nginx/certs/dhparam.pem;
        ssl_stapling on;
        ssl_stapling_verify on;
        ssl_trusted_certificate /etc/nginx/certs/fullchain.pem;

        # This is the CA, client certificates are checked against!
        # This CA is independent from the server certificate CA and could even be self-signed 
        ssl_client_certificate /etc/nginx/certs/ca.pem;
        # make verification optional, so we can display a 403 message to those
        # who fail authentication
        ssl_verify_client optional;

location / {
      # if the client-side certificate failed to authenticate, show a 403
      # message to the client
      if ($ssl_client_verify != SUCCESS) {
        return 403;
      }
		   set $app http://mtlsserver:8443;
                   proxy_pass $app;
}
```

This powers up an nginx listening at 443. All request without provided or invalid client certificate lead to an 403. Users can either use client certs via command line tools, or import the certs/key into browser (or better OS Key-Management). The browser will provide certs if asked by server (see point 3 [here](https://developers.cloudflare.com/access/static/mtls.png).

Also pomerium could benefit of this layer of extra security. Two options could be available:
1) global configuration: clients always need to provide client certificates, even to get the 301/2 redirect to the IDP
2) policy configuration: client certificates are needed per configured policy. IDP comes first, policy client certificate checks (if configured) next. Even different CAs for different policies could be provided here.

@desimone Does this make sense? I really like case 1. But I see case 2 more important.

**Comment by desimone** [comment]
Fixed with #751
